yarn-project/prover-client/src/prover/bb_prover.ts

```diff
@@ -280,27 +422,93 @@ export class BBNativeRollupProver implements CircuitProver {
     await this.verifyProof(circuitType, proof);
   }
 
-  public async getPublicKernelProof(
-    kernelRequest: PublicKernelNonTailRequest,
-  ): Promise<[PublicKernelCircuitPublicInputs, Proof]> {
-    const kernelOps = KernelArtifactMapping[kernelRequest.type];
-    if (kernelOps === undefined) {
-      throw new Error(`Unable to prove kernel type ${PublicKernelType[kernelRequest.type]}`);
+  /**
+   * Returns the verification key data for a circuit, will generate and cache it if not cached internally
+   * @param circuitType - The type of circuit for which the verification key is required
+   * @returns The verification key data
+   */
+  private async getVerificationKeyDataForCircuit(circuitType: ServerProtocolArtifact): Promise<VerificationKeyData> {
+    let promise = this.verificationKeys.get(circuitType);
+    if (!promise) {
+      promise = generateKeyForNoirCircuit(
+        this.config.bbBinaryPath,
+        this.config.bbWorkingDirectory,
+        circuitType,
+        ServerCircuitArtifacts[circuitType],
+        'vk',
+        logger.debug,
+      ).then(result => {
+        if (result.status === BB_RESULT.FAILURE) {
+          throw new Error(`Failed to generate verification key for ${circuitType}, ${result.reason}`);
+        }
+        return this.convertVk(result.vkPath!);
+      });
+      this.verificationKeys.set(circuitType, promise);
     }
-    const witnessMap = kernelOps.convertInputs(kernelRequest.inputs);
-
-    const [outputWitness, proof] = await this.createProof(witnessMap, kernelOps.artifact);
-
-    const result = kernelOps.convertOutputs(outputWitness);
-    return Promise.resolve([result, proof]);
+    return await promise;
   }
 
-  public async getPublicTailProof(kernelRequest: PublicKernelTailRequest): Promise<[KernelCircuitPublicInputs, Proof]> {
-    const witnessMap = convertPublicTailInputsToWitnessMap(kernelRequest.inputs);
+  /**
+   * Reads the verification key data stored at the specified location and parses into a VerificationKeyData
+   * @param filePath - The directory containing the verification key data files
+   * @returns The verification key data
+   */
+  private async convertVk(filePath: string): Promise<VerificationKeyData> {
+    const rawFields = await fs.readFile(`${filePath}/${VK_FIELDS_FILENAME}`, { encoding: 'utf-8' });
+    const rawBinary = await fs.readFile(`${filePath}/${VK_FILENAME}`);
+    const fieldsJson = JSON.parse(rawFields);
+    const fields = fieldsJson.map(Fr.fromString);
+    const vk: VerificationKeyData = {
+      hash: fields[VK_HASH_INDEX],
+      keyAsFields: fields.slice(1) as Tuple<Fr, typeof VERIFICATION_KEY_LENGTH_IN_FIELDS>,
```

Comment:
Why is this removing the first field? Is the first field the length and since the vk is fixed size we don't need it anymore?

Reply:
The fields version is specifically formatted for passing into a Noir circuit. The first field is the hash. The rest of the fields are the key. It just so happens that some of the first few fields provide some useful information which I have picked out. 

I have refactored this slightly which might make it a bit clear.